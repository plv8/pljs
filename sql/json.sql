CREATE FUNCTION valid_json(json text) RETURNS boolean
LANGUAGE pljs IMMUTABLE STRICT
AS $$
  try {
    JSON.parse(json);
    return true;
  } catch(e) {
    return false;
  }
$$;

SELECT valid_json('{"foo": "bar"}'::TEXT);

CREATE FUNCTION get_key(key text, json_data json) RETURNS json
LANGUAGE pljs IMMUTABLE STRICT
AS $$
  var val = json_data[key];
  var ret = {};
  ret[key] = val;
  return JSON.stringify(ret);
$$;

CREATE TABLE jsononly (
    data json
);

COPY jsononly (data) FROM stdin;
{"ok": true}
\.

-- Call twice to test the function cache.
SELECT get_key('ok', data) FROM jsononly;
SELECT get_key('ok', data) FROM jsononly;
