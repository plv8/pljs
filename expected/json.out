CREATE FUNCTION valid_json(json text) RETURNS boolean
LANGUAGE pljs IMMUTABLE STRICT
AS $$
  try {
    JSON.parse(json);
    return true;
  } catch(e) {
    return false;
  }
$$;
SELECT valid_json('{"foo": "bar"}'::TEXT);
 valid_json 
------------
 t
(1 row)

CREATE FUNCTION get_key(key text, json_data json) RETURNS json
LANGUAGE pljs IMMUTABLE STRICT
AS $$
  var val = json_data[key];
  var ret = {};
  ret[key] = val;
  return JSON.stringify(ret);
$$;
CREATE TABLE jsononly (
    data json
);
COPY jsononly (data) FROM stdin;
-- Call twice to test the function cache.
SELECT get_key('ok', data) FROM jsononly;
     get_key     
-----------------
 "{\"ok\":true}"
(1 row)

SELECT get_key('ok', data) FROM jsononly;
     get_key     
-----------------
 "{\"ok\":true}"
(1 row)

